### ISSUE TITLE
BUG: regex="(ab|cd)+c" doesn't match input="abc"

### ISSUE THREAD

**Issue by gogo9th** [mattbucknall/subreg#4]
The way to fix it is in function static int parse_repetition(state_t* state), change the for-loop block

```
   for (;;)
   {  
      state->regex = regex_begin;
      result = parse_literal(state);
      if ( is_bad_result(result) ) return result;
      
      if ( !is_match_result(result) ) break;
   }
```

to the following:

```
   for (;;)
   {  
      state->regex = regex_begin;
      const char* input_checkpoint = state->input;
      result = parse_literal(state);
      if ( is_bad_result(result) ) return result;
      
      if ( !is_match_result(result) ) { state->input = input_checkpoint; break;}
   }
```

**Comment by mattbucknall** [comment]
Fixed: 9b16f10e86b40516514db715f6f052c534f7b7ad

**Comment by gogo9th** [comment]
In commit https://github.com/mattbucknall/subreg/commit/a443963409b824c9767639e03e770a591e5777dc, we need to change:

```
    else if ( rc == '+' )
    {
        if ( !is_match_result(result) ) return SUBREG_RESULT_NO_MATCH;
    }
    else if ( rc == '*' )
    {
        if ( !is_match_result(result) )
        {
            state->regex++;
            state->input = check_point;
            return SUBREG_RESULT_INTERNAL_MATCH;
        }
}
```

to:

```
    else if ( rc == '+' )
    {
        if ( !is_match_result(result) || state->input == check_point ) return SUBREG_RESULT_NO_MATCH;
    }
    else if ( rc == '*' )
    {
        if ( !is_match_result(result) || state->input == check_point )
        {
            state->regex++;
            state->input = check_point;
            return SUBREG_RESULT_INTERNAL_MATCH;
        }
}
```

Otherwise, we can fall into an infinite loop.
